File: engine/src/core/window.c
/**
 * @file window.c
 * @brief Window management implementation for Cardinal Engine
 *
 * This file implements cross-platform window management functionality using
 * GLFW. It handles window creation, event processing, input handling, and
 * integration with the Vulkan rendering system.
 *
 * Key features:
 * - Cross-platform window creation (Windows, Linux, macOS)
 * - Vulkan surface integration
 * - Input event handling (keyboard, mouse)
 * - Window resize and framebuffer callbacks
 * - Fullscreen and windowed mode support
 * - Error handling and logging integration
 *
 * Platform-specific features:
 * - Windows: Native Win32 handle access
 * - Linux: Wayland compatibility
 * - macOS: Metal surface support preparation
 *
 * The implementation provides a clean abstraction over GLFW while
 * maintaining access to platform-specific functionality when needed
 * for advanced rendering features.
 *
 * @author Markus Maurer
 * @version 1.0
 */

#include <GLFW/glfw3.h>
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#ifdef _WIN32
    #define GLFW_EXPOSE_NATIVE_WIN32
    #include <GLFW/glfw3native.h>
#endif

#include "cardinal/core/log.h"
#include "cardinal/core/window.h"

/**
 * @brief GLFW error callback function.
 *
 * Logs GLFW errors.
 *
 * @param error Error code.
 * @param desc Error description.
 *
 * @todo Integrate with centralized error handling system.
 */
static void glfw_error_callback(int error, const char* desc) {
    CARDINAL_LOG_ERROR("GLFW error %d: %s", error, desc ? desc : "(null)");
}

static void framebuffer_resize_callback(GLFWwindow* window, int width, int height) {
    CardinalWindow* win = (CardinalWindow*)glfwGetWindowUserPointer(window);
    if (!win)
        return;
    if (width > 0 && height > 0) {
        win->resize_pending = true;
        win->new_width = (uint32_t)width;
        win->new_height = (uint32_t)height;
        win->width = (uint32_t)width;
        win->height = (uint32_t)height;
        if (win->resize_callback) {
            win->resize_callback(win->new_width, win->new_height, win->resize_user_data);
        }
    }
}

static void window_iconify_callback(GLFWwindow* window, int iconified) {
    CardinalWindow* win = (CardinalWindow*)glfwGetWindowUserPointer(window);
    if (!win)
        return;
    win->was_minimized = win->is_minimized;
    win->is_minimized = (iconified == GLFW_TRUE);
    if (win->was_minimized && !win->is_minimized) {
        int w = 0, h = 0;
        glfwGetFramebufferSize(window, &w, &h);
        if (w > 0 && h > 0) {
            win->resize_pending = true;
            win->new_width = (uint32_t)w;
            win->new_height = (uint32_t)h;
            win->width = (uint32_t)w;
            win->height = (uint32_t)h;
            if (win->resize_callback) {
                win->resize_callback(win->new_width, win->new_height, win->resize_user_data);
            }
        }
    }
}
/**
 * @brief Creates a new window with the given configuration.
 *
 * @param config Window configuration.
 * @return Pointer to the created window or NULL on failure.
 *
 * @todo Refactor to support multiple window creation.
 * @todo Integrate Vulkan surface creation directly for better renderer compatibility.
 * @todo Support multiple monitors and fullscreen modes.
 * @todo Add validation for config parameters.
 */
CardinalWindow* cardinal_window_create(const CardinalWindowConfig* config) {
    assert(config);
    CARDINAL_LOG_INFO("cardinal_window_create: begin");
    if (!glfwInit()) {
        CARDINAL_LOG_ERROR("GLFW init failed");
        return NULL;
    }

    glfwSetErrorCallback(glfw_error_callback);

    glfwWindowHint(GLFW_CLIENT_API, GLFW_NO_API);
    glfwWindowHint(GLFW_RESIZABLE, config->resizable ? GLFW_TRUE : GLFW_FALSE);

    GLFWwindow* handle =
        glfwCreateWindow((int)config->width, (int)config->height, config->title, NULL, NULL);
    if (!handle) {
        CARDINAL_LOG_ERROR("GLFW create window failed");
        glfwTerminate();
        return NULL;
    }

    CardinalWindow* win = (CardinalWindow*)calloc(1, sizeof(CardinalWindow));
    win->handle = handle;
    win->width = config->width;
    win->height = config->height;
    win->should_close = false;
    // Register GLFW callbacks for resize/minimize
    glfwSetWindowUserPointer(handle, win);
    glfwSetFramebufferSizeCallback(handle, framebuffer_resize_callback);
    glfwSetWindowIconifyCallback(handle, window_iconify_callback);

    // Initialize mutex for thread safety
#ifdef _WIN32
    InitializeCriticalSection(&win->mutex);
#else
    if (pthread_mutex_init(&win->mutex, NULL) != 0) {
        CARDINAL_LOG_ERROR("Failed to initialize window mutex");
        glfwDestroyWindow(win->handle);
        free(win);
        return NULL;
    }
#endif

    CARDINAL_LOG_INFO("cardinal_window_create: success");
    return win;
}

/**
 * @brief Polls for window events.
 *
 * @param window The window to poll.
 *
 * @todo Implement custom event queue for better control.
 * @todo Improve by adding custom event dispatching system.
 * @todo Add support for touch and gesture events.
 */
void cardinal_window_poll(CardinalWindow* window) {
    if (!window)
        return;

    // Lock mutex for thread-safe polling
#ifdef _WIN32
    EnterCriticalSection(&window->mutex);
#else
    pthread_mutex_lock(&window->mutex);
#endif

    glfwPollEvents();
    window->should_close = glfwWindowShouldClose(window->handle) != 0;

    // Unlock mutex
#ifdef _WIN32
    LeaveCriticalSection(&window->mutex);
#else
    pthread_mutex_unlock(&window->mutex);
#endif
}

/**
 * @brief Checks if the window should close.
 *
 * @param window The window to check.
 * @return true if should close, false otherwise.
 *
 * @todo Add custom close conditions (e.g., based on application state).
 */
bool cardinal_window_should_close(const CardinalWindow* window) {
    return window ? window->should_close : true;
}

/**
 * @brief Destroys the window and cleans up resources.
 * @param window The window to destroy.
 *
 * @todo Ensure proper cleanup of associated Vulkan resources.
 * @todo Add support for Vulkan extension VK_KHR_portability_subset for better cross-platform
 * @todo Ensure thread-safe destruction.
 * @todo Clean up associated input callbacks.
 * handling.
 */
void cardinal_window_destroy(CardinalWindow* window) {
    if (!window)
        return;

    // Lock mutex for thread-safe destruction
#ifdef _WIN32
    EnterCriticalSection(&window->mutex);
#else
    pthread_mutex_lock(&window->mutex);
#endif

    if (window->handle) {
        glfwDestroyWindow(window->handle);
        window->handle = NULL;
    }
    glfwTerminate();

    // Unlock and destroy mutex
#ifdef _WIN32
    LeaveCriticalSection(&window->mutex);
    DeleteCriticalSection(&window->mutex);
#else
    pthread_mutex_unlock(&window->mutex);
    pthread_mutex_destroy(&window->mutex);
#endif

    free(window);
}

void* cardinal_window_get_native_handle(const CardinalWindow* window) {
    if (!window || !window->handle)
        return NULL;
#ifdef _WIN32
    return (void*)glfwGetWin32Window(window->handle);
#else
    return NULL;
#endif
}

bool cardinal_window_is_minimized(const CardinalWindow* window) {
    return window ? window->is_minimized : false;
}

File: engine/src/core/window.zig
//! Window creation and input callback routing.
//!
//! Wraps GLFW window handles behind a C-ABI `CardinalWindow` and allows clients to register
//! callbacks for resize, key, mouse button, and cursor movement.
const std = @import("std");
const log = @import("log.zig");
const builtin = @import("builtin");
const platform = @import("platform.zig");

const win_log = log.ScopedLogger("WINDOW");

const c = platform.c;

var g_glfw_lock = std.Thread.Mutex{};
var g_glfw_initialized = false;
var g_glfw_window_count: u32 = 0;

fn glfw_acquire() bool {
    g_glfw_lock.lock();
    defer g_glfw_lock.unlock();

    if (!g_glfw_initialized) {
        if (c.glfwInit() == c.GLFW_FALSE) {
            win_log.err("GLFW init failed", .{});
            return false;
        }
        _ = c.glfwSetErrorCallback(glfw_error_callback);
        g_glfw_initialized = true;
    }

    g_glfw_window_count += 1;
    return true;
}

fn glfw_release() void {
    g_glfw_lock.lock();
    defer g_glfw_lock.unlock();

    if (g_glfw_window_count > 0) {
        g_glfw_window_count -= 1;
    }

    if (g_glfw_window_count == 0 and g_glfw_initialized) {
        c.glfwTerminate();
        g_glfw_initialized = false;
    }
}

/// Configuration used by `cardinal_window_create`.
pub const CardinalWindowConfig = extern struct {
    title: [*:0]const u8,
    width: u32,
    height: u32,
    resizable: bool,
};

/// Opaque window wrapper around a GLFW window handle plus per-window callbacks/state.
pub const CardinalWindow = struct {
    handle: ?*c.GLFWwindow,
    width: u32,
    height: u32,
    should_close: bool,
    mutex: std.Thread.Mutex,

    resize_pending: bool,
    new_width: u32,
    new_height: u32,
    is_minimized: bool,
    was_minimized: bool,

    resize_callback: ?*const fn (u32, u32, ?*anyopaque) callconv(.c) void,
    resize_user_data: ?*anyopaque,

    key_callback: ?*const fn (c_int, c_int, c_int, c_int, ?*anyopaque) callconv(.c) void,
    key_user_data: ?*anyopaque,

    mouse_button_callback: ?*const fn (c_int, c_int, c_int, ?*anyopaque) callconv(.c) void,
    mouse_button_user_data: ?*anyopaque,

    cursor_pos_callback: ?*const fn (f64, f64, ?*anyopaque) callconv(.c) void,
    cursor_pos_user_data: ?*anyopaque,

    content_scale_x: f32,
    content_scale_y: f32,
};

/// Updates content scale state when the OS DPI scale changes.
fn window_content_scale_callback(window: ?*c.GLFWwindow, xscale: f32, yscale: f32) callconv(.c) void {
    const win = @as(?*CardinalWindow, @ptrCast(@alignCast(c.glfwGetWindowUserPointer(window))));
    if (win) |w| {
        w.content_scale_x = xscale;
        w.content_scale_y = yscale;
    }
}

/// Forwards GLFW key events to the registered callback.
fn key_callback(window: ?*c.GLFWwindow, key: c_int, scancode: c_int, action: c_int, mods: c_int) callconv(.c) void {
    const win = @as(?*CardinalWindow, @ptrCast(@alignCast(c.glfwGetWindowUserPointer(window))));
    if (win) |w| {
        if (w.key_callback) |cb| {
            cb(key, scancode, action, mods, w.key_user_data);
        }
    }
}

/// Forwards GLFW mouse button events to the registered callback.
fn mouse_button_callback(window: ?*c.GLFWwindow, button: c_int, action: c_int, mods: c_int) callconv(.c) void {
    const win = @as(?*CardinalWindow, @ptrCast(@alignCast(c.glfwGetWindowUserPointer(window))));
    if (win) |w| {
        if (w.mouse_button_callback) |cb| {
            cb(button, action, mods, w.mouse_button_user_data);
        }
    }
}

/// Forwards GLFW cursor position events to the registered callback.
fn cursor_pos_callback(window: ?*c.GLFWwindow, xpos: f64, ypos: f64) callconv(.c) void {
    const win = @as(?*CardinalWindow, @ptrCast(@alignCast(c.glfwGetWindowUserPointer(window))));
    if (win) |w| {
        if (w.cursor_pos_callback) |cb| {
            cb(xpos, ypos, w.cursor_pos_user_data);
        }
    }
}

/// GLFW error callback.
fn glfw_error_callback(error_code: c_int, description: [*c]const u8) callconv(.c) void {
    win_log.err("GLFW error {d}: {s}", .{ error_code, if (description != null) std.mem.span(description) else "(null)" });
}

/// Forwards framebuffer resize events and marks `resize_pending`.
fn framebuffer_resize_callback(window: ?*c.GLFWwindow, width: c_int, height: c_int) callconv(.c) void {
    const win = @as(?*CardinalWindow, @ptrCast(@alignCast(c.glfwGetWindowUserPointer(window))));
    if (win) |w| {
        if (width > 0 and height > 0) {
            w.resize_pending = true;
            w.new_width = @intCast(width);
            w.new_height = @intCast(height);
            w.width = @intCast(width);
            w.height = @intCast(height);
            if (w.resize_callback) |cb| {
                cb(w.new_width, w.new_height, w.resize_user_data);
            }
        }
    }
}

/// Tracks minimize/restore and triggers a resize callback on restore.
fn window_iconify_callback(window: ?*c.GLFWwindow, iconified: c_int) callconv(.c) void {
    const win = @as(?*CardinalWindow, @ptrCast(@alignCast(c.glfwGetWindowUserPointer(window))));
    if (win) |w| {
        w.was_minimized = w.is_minimized;
        w.is_minimized = (iconified == c.GLFW_TRUE);

        if (w.was_minimized and !w.is_minimized) {
            var width: c_int = 0;
            var height: c_int = 0;
            c.glfwGetFramebufferSize(window, &width, &height);
            if (width > 0 and height > 0) {
                w.resize_pending = true;
                w.new_width = @intCast(width);
                w.new_height = @intCast(height);
                w.width = @intCast(width);
                w.height = @intCast(height);
                if (w.resize_callback) |cb| {
                    cb(w.new_width, w.new_height, w.resize_user_data);
                }
            }
        }
    }
}

/// Creates a window and initializes GLFW for Vulkan usage.
pub export fn cardinal_window_create(config: *const CardinalWindowConfig) callconv(.c) ?*CardinalWindow {
    win_log.info("cardinal_window_create: begin", .{});

    if (!glfw_acquire()) {
        return null;
    }

    c.glfwWindowHint(c.GLFW_CLIENT_API, c.GLFW_NO_API);
    c.glfwWindowHint(c.GLFW_RESIZABLE, if (config.resizable) c.GLFW_TRUE else c.GLFW_FALSE);

    const handle = c.glfwCreateWindow(@intCast(config.width), @intCast(config.height), config.title, null, null);
    if (handle == null) {
        win_log.err("GLFW create window failed", .{});
        glfw_release();
        return null;
    }

    const win_ptr = c.malloc(@sizeOf(CardinalWindow));
    if (win_ptr == null) {
        c.glfwDestroyWindow(handle);
        glfw_release();
        return null;
    }
    const win = @as(*CardinalWindow, @ptrCast(@alignCast(win_ptr)));
    win.* = std.mem.zeroes(CardinalWindow);

    win.handle = handle;
    win.width = config.width;
    win.height = config.height;
    win.should_close = false;

    var x_scale: f32 = 1.0;
    var y_scale: f32 = 1.0;
    c.glfwGetWindowContentScale(handle, &x_scale, &y_scale);
    win.content_scale_x = x_scale;
    win.content_scale_y = y_scale;

    c.glfwSetWindowUserPointer(handle, win);
    _ = c.glfwSetFramebufferSizeCallback(handle, framebuffer_resize_callback);
    _ = c.glfwSetWindowContentScaleCallback(handle, window_content_scale_callback);
    _ = c.glfwSetWindowIconifyCallback(handle, window_iconify_callback);
    _ = c.glfwSetKeyCallback(handle, key_callback);
    _ = c.glfwSetMouseButtonCallback(handle, mouse_button_callback);
    _ = c.glfwSetCursorPosCallback(handle, cursor_pos_callback);

    win.mutex = .{};

    win_log.info("cardinal_window_create: success", .{});
    return win;
}

/// Polls the window event loop and updates `should_close`.
pub export fn cardinal_window_poll(window: ?*CardinalWindow) callconv(.c) void {
    if (window) |win| {
        win.mutex.lock();
        c.glfwPollEvents();
        win.should_close = (c.glfwWindowShouldClose(win.handle) != 0);

        win.mutex.unlock();
    }
}

/// Returns true if the window has requested close.
pub export fn cardinal_window_should_close(window: ?*const CardinalWindow) callconv(.c) bool {
    if (window) |win| {
        return win.should_close;
    }
    return true;
}

/// Destroys the window.
pub export fn cardinal_window_destroy(window: ?*CardinalWindow) callconv(.c) void {
    if (window) |win| {
        win.mutex.lock();

        if (win.handle) |h| {
            c.glfwDestroyWindow(h);
            win.handle = null;
        }

        win.mutex.unlock();

        c.free(win);
        glfw_release();
    }
}

/// Returns the underlying platform native window handle (if available).
pub export fn cardinal_window_get_native_handle(window: ?*const CardinalWindow) callconv(.c) ?*anyopaque {
    if (window) |win| {
        if (win.handle) |h| {
            if (builtin.os.tag == .windows) {
                return c.glfwGetWin32Window(h);
            }
        }
    }
    return null;
}

pub export fn cardinal_window_is_minimized(window: ?*const CardinalWindow) callconv(.c) bool {
    if (window) |win| {
        return win.is_minimized;
    }
    return false;
}

pub export fn cardinal_window_set_key_callback(window: ?*CardinalWindow, callback: ?*const fn (c_int, c_int, c_int, c_int, ?*anyopaque) callconv(.c) void, user_data: ?*anyopaque) callconv(.c) void {
    if (window) |win| {
        win.key_callback = callback;
        win.key_user_data = user_data;
    }
}

pub export fn cardinal_window_set_mouse_button_callback(window: ?*CardinalWindow, callback: ?*const fn (c_int, c_int, c_int, ?*anyopaque) callconv(.c) void, user_data: ?*anyopaque) callconv(.c) void {
    if (window) |win| {
        win.mouse_button_callback = callback;
        win.mouse_button_user_data = user_data;
    }
}

pub export fn cardinal_window_set_cursor_pos_callback(window: ?*CardinalWindow, callback: ?*const fn (f64, f64, ?*anyopaque) callconv(.c) void, user_data: ?*anyopaque) callconv(.c) void {
    if (window) |win| {
        win.cursor_pos_callback = callback;
        win.cursor_pos_user_data = user_data;
    }
}

pub export fn cardinal_window_get_content_scale(window: ?*const CardinalWindow, x_scale: ?*f32, y_scale: ?*f32) callconv(.c) void {
    if (window) |win| {
        if (x_scale) |x| x.* = win.content_scale_x;
        if (y_scale) |y| y.* = win.content_scale_y;
    } else {
        if (x_scale) |x| x.* = 1.0;
        if (y_scale) |y| y.* = 1.0;
    }
}

// Additional window functions for Project Manager
pub export fn cardinal_window_set_size(window: ?*CardinalWindow, width: u32, height: u32) callconv(.c) void {
    if (window) |win| {
        if (win.handle) |h| {
            c.glfwSetWindowSize(h, @intCast(width), @intCast(height));
            win.width = width;
            win.height = height;
        }
    }
}

pub export fn cardinal_window_set_title(window: ?*CardinalWindow, title: [*c]const u8) callconv(.c) void {
    if (window) |win| {
        if (win.handle) |h| {
            c.glfwSetWindowTitle(h, title);
        }
    }
}

pub export fn cardinal_window_maximize(window: ?*CardinalWindow) callconv(.c) void {
    if (window) |win| {
        if (win.handle) |h| {
            c.glfwMaximizeWindow(h);
        }
    }
}

pub export fn cardinal_window_center(window: ?*CardinalWindow) callconv(.c) void {
    if (window) |win| {
        if (win.handle) |h| {
            const monitor = c.glfwGetPrimaryMonitor();
            if (monitor) |m| {
                const mode = c.glfwGetVideoMode(m);
                if (mode) |vmode| {
                    var width: c_int = 0;
                    var height: c_int = 0;
                    c.glfwGetWindowSize(h, &width, &height);

                    const x = @divTrunc(vmode.*.width - width, 2);
                    const y = @divTrunc(vmode.*.height - height, 2);
                    c.glfwSetWindowPos(h, x, y);
                }
            }
        }
    }
}

pub export fn cardinal_window_restore(window: ?*CardinalWindow) callconv(.c) void {
    if (window) |win| {
        if (win.handle) |h| {
            c.glfwRestoreWindow(h);
        }
    }
}

pub export fn cardinal_window_get_glfw_handle(window: ?*const CardinalWindow) callconv(.c) ?*anyopaque {
    if (window) |win| {
        return win.handle;
    }
    return null;
}
